src/hidpp20.c

```diff
@@ -2412,91 +2421,61 @@ hidpp20_onboard_profiles_parse(struct hidpp20_device *device,
 	crc_valid = hidpp20_onboard_profiles_is_sector_valid(device,
 							     profiles->sector_size,
 							     data);
-	if (crc_valid || !check_crc) {
+	if (crc_valid) {
 		for (i = 0; i < profiles->num_profiles; i++) {
 			uint8_t *d = data + 4 * i;
 
-			if (d[0] == 0xFF && d[1] == 0xFF)
+			if (d[HIDPP20_PROFILE_DIR_END1] == HIDPP20_PROFILE_DIR_END_VALUE &&
+			    d[HIDPP20_PROFILE_DIR_END2] == HIDPP20_PROFILE_DIR_END_VALUE)
 				break;
 
 			profiles->profiles[i].address = hidpp_get_unaligned_be_u16(d);
-			if (profiles->profiles[i].address != (dict_address | (i + 1)))
+
+			/* profile address sanity check */
+			if (profiles->profiles[i].address != (HIDPP20_USER_PROFILES_G402 | (i + 1)))
 				hidpp_log_info(&device->base,
-						"profile %d: error in the address: 0x%04x instead of 0x%04x\n",
-						i + 1,
-						profiles->profiles[i].address,
-						dict_address | (i + 1));
-			profiles->profiles[i].enabled = !!d[2];
-		}
-	}
+					       "profile %d: error in the address: 0x%04x instead of 0x%04x\n",
+					       i + 1,
+					       profiles->profiles[i].address,
+					       HIDPP20_USER_PROFILES_G402 | (i + 1));
 
-	/*
-	 * Even if the crc is wrong, we still fetch all the profiles, so as
-	 * to have the current stored values if some profiles are legitimate.
-	 */
-	 for (i = 0; i < profiles->num_profiles; i++) {
-		if (profiles->profiles[i].address == 0x0000) {
-			switch (dict_address) {
-			case 0x0000:
-				profiles->profiles[i].address = i + 1;
-				break;
-			case 0x0100:
-				if (i >= profiles->num_rom_profiles) {
-					profiles->profiles[i].address = dict_address | profiles->num_rom_profiles;
-				} else {
-					profiles->profiles[i].address = dict_address | (i + 1);
-				}
-				break;
-			default:
-				hidpp_log_error(&device->base,
-						"this should never happen: dict address is 0x%04x\n",
-						dict_address);
-				return -EINVAL;
-			}
+			profiles->profiles[i].enabled = !!d[HIDPP20_PROFILE_DIR_ENABLED];
```

Comment:
Didn't you said that we should not try to read the disabled profiles?

Reply:
Yeah but nevermind. We should still read disabled profiles. If a user wishes to enable the profile we need to have the data there.

We can either read them here or when they are enabled, I would prefer to read them here.